src/waggle/data/vision.py

```diff
@@ -135,34 +194,87 @@ def __enter__(self):
                 raise RuntimeError(
                     f"unable to open video capture for device {self.device!r}"
                 )
+            # spin up a thread to keep up with the camera frame rate
+            if self.enable_daemon:
+                self.daemon.start()
         self.context_depth += 1
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.context_depth -= 1
         if self.context_depth == 0:
+            if self.enable_daemon:
+                self.daemon_need_to_stop = True
+                self.daemon.join()
             self.capture.release()
-
-    def snapshot(self):
-        ok = self.capture.grab()
-        if not ok:
-            raise RuntimeError("failed to take snapshot")
-        timestamp = get_timestamp()
-        ok, data = self.capture.retrieve()
-        if not ok:
-            raise RuntimeError("failed to retrieve the taken snapshot")
-        return ImageSample(data=data, timestamp=timestamp, format=self.format)
-
-    def stream(self):
-        while True:
+    
+    def _run(self):
+        # we sleep slighly shorter than FPS to drain the buffer efficiently
+        fps = self.capture.get(cv2.CAP_PROP_FPS)
+        sleep = 0.01
+        if fps > 0 and fps < 100:
+            sleep = 1 / (fps + 1)
+        logging.debug(f'camera FPS is {fps}. the background thread sleeps {sleep} seconds in between grab()')
+        while not self.daemon_need_to_stop:
+            self.lock.acquire()
             ok = self.capture.grab()
             if not ok:
-                break
+                pass
```

Comment:
I think 99% of the time `.grab()` should succeed, and the 1% would be caused by a fatal error rather than a small error we can recover from. So, I think we should raise an RuntimeError when it happens.

Reply:
Sure. I think it's mostly a matter of how to signal that to the main thread. (For example, we _might_ just use the timeout when the main thread tries to acquire the lock.)

Reply:
Just trying to understand... Are you saying that the producer thread might silently die (or retry)? and the main thread (i.e., consumer function calls) gets timed out when it happens?

Reply:
The main thread will raise a RuntimeError if the background thread couldn't set the flag within 10 seconds (probably because the BG couldn't obtain a frame by the timeout). This way the end user will be notified by the main thread's RuntimeError.

```python
if not self._ready_for_next_frame.wait(timeout=10.):
    raise RuntimeError("failed to grab a frame from the background thread: timed out")
```